packages/datagrid/src/selectionmodel.ts

```diff
@@ -0,0 +1,337 @@
+    this._selectionMode = value;
+
+    // Clear the current selections.
+    this.clear();
+  }
+
+  /**
+   * Test whether any selection intersects a row.
+   *
+   * @param index - The row index of interest.
+   *
+   * @returns Whether any selection intersects the row.
+   *
+   * #### Notes
+   * This method may be reimplemented in a subclass.
+   */
+  isRowSelected(index: number): boolean {
+    return some(this.selections(), s => Private.containsRow(s, index));
+  }
+
+  /**
+   * Test whether any selection intersects a column.
+   *
+   * @param index - The column index of interest.
+   *
+   * @returns Whether any selection intersects the column.
+   *
+   * #### Notes
+   * This method may be reimplemented in a subclass.
+   */
+  isColumnSelected(index: number): boolean {
+    return some(this.selections(), s => Private.containsColumn(s, index));
+  }
+
+  /**
+   * Test whether any selection intersects a cell.
+   *
+   * @param row - The row index of interest.
+   *
+   * @param column - The column index of interest.
+   *
+   * @returns Whether any selection intersects the cell.
+   *
+   * #### Notes
+   * This method may be reimplemented in a subclass.
+   */
+  isCellSelected(row: number, column: number): boolean {
+    return some(this.selections(), s => Private.containsCell(s, row, column));
+  }
+
+  /**
+   * Emit the `changed` signal for the selection model.
+   *
+   * #### Notes
+   * Subclasses should call this method whenever the selection model
+   * has changed so that attached data grids can update themselves.
+   */
+  protected emitChanged(): void {
+    this._changed.emit(undefined);
+  }
+
+  private _changed = new Signal<this, void>(this);
+  private _selectionMode: SelectionModel.SelectionMode = 'cell';
+}
+
+
+/**
+ * The namespace for the `SelectionModel` class statics.
+ */
+export
+namespace SelectionModel {
+  /**
+   * A type alias for the selection mode.
+   */
+  export
+  type SelectionMode = 'row' | 'column' | 'cell';
+
+  /**
+   * A type alias for the clear mode.
+   */
+  export
+  type ClearMode = 'all' | 'current' | 'none';
+
+  /**
+   * A type alias for the select args.
+   */
+  export
+  type SelectArgs = {
+    /**
+     * The first row of the selection.
+     */
+    r1: number;
+
+    /**
+     * The first column of the selection.
+     */
+    c1: number;
+
+    /**
+     * The second row of the selection.
+     */
+    r2: number;
+
+    /**
+     * The second column of the selection.
+     */
+    c2: number;
+
+    /**
+     * The row index for the cursor.
+     */
+    cursorRow: number;
+
+    /**
+     * The column index for the cursor.
+     */
+    cursorColumn: number;
+
+    /**
+     * Which of the existing selections to clear.
+     */
+    clear: ClearMode;
+  };
+
+  /**
+   * A type alias for a selection in a selection model.
+   */
+  export
+  type Selection = {
+    /**
+     * The first row of the selection.
+     */
+    readonly r1: number;
```

Comment:
Is the reason for `r1`, `c1`, `r2`, `c2` instead of `top`, `left`, `bottom`, `right` because you want the direction of selection to not be implied (*e.g.*, I dragged the selection from bottom up to top instead of vice versa)?

Reply:
Exactly. It's to indicate the direction of the selection. `r1` is not necessarily less than `r2`. It's also less to type than `startRow` and `endRow`.

However, it's the first place in the whole package where I've used abbreviations like that, so I could be convinced to use `startRow` and `endRow` or something similar.

Reply:
I had to read what r2 and c2 meant twice because it wasn't clear they were end points.  I think r0 and r1 and calling them start and end in the comments would be clearer.